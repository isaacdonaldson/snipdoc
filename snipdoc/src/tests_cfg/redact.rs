#[cfg(windows)]
pub const REGEX_REPLACE_LINE_ENDING: &str = r"\\r\\n";

#[cfg(not(windows))]
pub const REGEX_REPLACE_LINE_ENDING: &str = r"\\n";

#[must_use]
pub fn all() -> Vec<(&'static str, &'static str)> {
    vec![(REGEX_REPLACE_LINE_ENDING, "[NEW_LINE]")]
}
