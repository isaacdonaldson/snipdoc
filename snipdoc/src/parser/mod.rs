pub mod collector;
pub mod html_tag;
pub mod injector;

use std::collections::HashMap;

use pest::Parser;
// use pest::{iterators::Pairs, Parser};
use pest_derive::Parser;

use crate::{db::Snippet, errors::ParserResult};

#[derive(Parser)]
#[grammar = "snippet.pest"]
pub struct SnippetParse;

/// A structure representing a file to be parsed.
pub struct ParseFile<'a> {
    pub input: &'a str,
}

impl<'a> ParseFile<'a> {
    /// Constructs a new [`ParseFile`] with the provided input.
    #[must_use]
    pub const fn new(input: &'a str) -> Self {
        Self { input }
    }

    // /// Constructs a new [`ParseFile`] with the provided input.
    // #[must_use]
    // pub const fn new(input: &'a str) -> Self {
    //     Self { input }
    // }

    /// Parses the input file content and extracts snippets.
    ///
    /// # Errors
    ///
    /// This function may return an error if it fails to parse the input file.
    /// Other errors encountered during parsing will be logged.
    pub fn parse(&self) -> ParserResult<'_, Vec<collector::CollectSnippet>> {
        let pairs = SnippetParse::parse(Rule::file, self.input)?;

        let mut findings: Vec<collector::CollectSnippet> = vec![];
        collector::collect_snippets(pairs, &mut findings);
        Ok(findings)
    }

    /// Injects snippets in the input file content based on the provided
    /// [`Snippet`] map.
    ///
    /// # Errors
    ///
    /// This function may return an error if it fails to parse the input file.
    /// Other errors encountered during parsing will be logged.
    pub fn inject(
        &self,
        snippets: &HashMap<String, &Snippet>,
    ) -> ParserResult<'_, injector::InjectSummary> {
        let pairs = SnippetParse::parse(Rule::file, self.input)?;

        let mut inject_summary = injector::InjectSummary::default();
        injector::inject_snippets(pairs, &mut inject_summary, snippets)?;

        Ok(inject_summary)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;
    use crate::tests_cfg;

    #[test]
    fn can_collect() {
        let content = r#"# Snipdoc
# parse the snippet with `inject_from` attribute
<!-- <snip id="description" inject_from="code"> -->
<!-- </snip> -->

# parse snippet with html comment
<!-- <snip id="installation"> -->
$ cargo install snipdoc
$ ssnipdoc --version
<!-- </snip> -->

# parse snippet without any spaces between the comment 
<!--<snip id="no-spaces">-->
$ cargo install snipdoc
$ ssnipdoc --version
<!--</snip>-->

# parse snippet with double slash comment
// <snip id="double-slash"> 
double-slash
// </snip> 

# parse snippet with triple slash comment
/// <snip id="triple-slash"> 
triple-slash
/// </snip> 
/// 
# parse snippet with triple hashtag comment
# <snip id="hashtag"> 
hashtag
# </snip> 

# Inner snippets
<!-- <snip id="level-1" -->
Level 1
// <snip id="level-2"> 
Level 2
// <snip id="level-3"> 
Level 3
// </snip>
// </snip>
<!-- </snip> -->
"#;

        let parser = ParseFile::new(content);
        assert_debug_snapshot!(parser.parse());
    }

    #[test]
    fn parsing_error() {
        let content = r#"<!-- <snip id="description"></snip"#;
        let parser: ParseFile = ParseFile::new(content);
        assert!(parser.parse().is_err());
    }

    #[test]
    fn get_inject() {
        let content = r#"# Snipdoc

<!-- <snip id="installation" inject_from="code"> -->
# inject `installation` snippet id from code snippet kind
<!-- </snip> -->

<!-- <snip id="inject_from_yaml" inject_from="yaml"> -->
# inject `inject_from_yaml` snippet id from yaml snippet kind
<!-- </snip> -->

<!-- <snip id="inject_from_yaml" inject_from="code"> -->
# Skip injection, `inject_from_yaml` snippet id not exists in code
<!-- </snip> -->

<!-- <snip id="inject_from_yaml" inject_from="any"> -->
# inject_from is any, and this id exists in the yaml
<!-- </snip> -->

<!-- <snip id="description" inject_from="code" add_prefix="//! "> -->
# Adding the prefix for each line
<!-- </snip> -->

<!-- <snip id="description" inject_from="code" strip_prefix="snip"> -->
# Strip `snip` word from prefix for each line
<!-- </snip> -->

<!-- <snip id="description" inject_from="code" template="```sh\n{snippet}\n```"> -->
# Add template to inject snippet
<!-- </snip> -->

<!-- <snip id="description" inject_from="code"> -->
snipdoc
<!-- </snip> -->

<!-- <snip id="not-found" inject_from="code"> -->
not-found
<!-- </snip> -->

"#;
        let parser: ParseFile = ParseFile::new(content);
        let snippets: HashMap<String, Snippet> = tests_cfg::get_snippet_to_inject();
        let snippet_refs: HashMap<String, &Snippet> =
            snippets.iter().map(|(k, v)| (k.clone(), v)).collect();

        assert_debug_snapshot!(parser.inject(&snippet_refs));
    }
}
