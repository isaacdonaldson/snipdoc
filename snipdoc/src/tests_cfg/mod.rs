use std::{collections::BTreeMap, path::PathBuf};
pub mod redact;
use crate::parser::{collector::CollectSnippet, Snippet, SnippetKind, SnippetTemplate};

#[must_use]
pub fn get_collect_snippets() -> Vec<CollectSnippet> {
    vec![
        CollectSnippet {
            id: "description".to_string(),
            snippet: vec!["test".to_string(), "snipdoc".to_string()],
            inject_from: None,
            tag_open: "<snip id=\"description\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        },
        CollectSnippet {
            id: "installation".to_string(),
            snippet: vec![
                "```".to_string(),
                "cargo install snipdoc".to_string(),
                "```".to_string(),
            ],
            inject_from: None,
            tag_open: "<snip id=\"install\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        },
        CollectSnippet {
            id: "from-yaml".to_string(),
            snippet: vec!["ignore-snippet".to_string()],
            inject_from: Some(SnippetKind::Yaml),
            tag_open: "<snip id=\"from-yaml\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        },
    ]
}

#[must_use]
pub fn get_custom_templates() -> BTreeMap<String, SnippetTemplate> {
    BTreeMap::from([
        (
            "CUSTOM_ID_1".to_string(),
            SnippetTemplate {
                content: "---> {snippet}".to_string(),
            },
        ),
        (
            "CUSTOM_ID_2".to_string(),
            SnippetTemplate {
                content: "({snippet})".to_string(),
            },
        ),
    ])
}

#[must_use]
pub fn get_snippet() -> Snippet {
    Snippet {
        id: "test".to_string(),
        content: "$ cargo install snipdoc\n$ snipdoc --version".to_string(),
        kind: SnippetKind::Code,
        path: PathBuf::from("main.rs"),
    }
}

#[must_use]
pub fn get_snippet_to_inject() -> BTreeMap<String, Snippet> {
    BTreeMap::from([
        (
            "description".to_string(),
            Snippet {
                id: "test".to_string(),
                content: "snipdoc".to_string(),
                kind: SnippetKind::Code,
                path: PathBuf::from("main.rs"),
            },
        ),
        (
            "installation".to_string(),
            Snippet {
                id: "test".to_string(),
                content: "$ cargo install snipdoc\n$ snipdoc --version".to_string(),
                kind: SnippetKind::Code,
                path: PathBuf::from("main.rs"),
            },
        ),
        (
            "inject_from_yaml".to_string(),
            Snippet {
                id: "test".to_string(),
                content: "inject_from_yaml".to_string(),
                kind: SnippetKind::Yaml,
                path: PathBuf::from("main.rs"),
            },
        ),
    ])
}
