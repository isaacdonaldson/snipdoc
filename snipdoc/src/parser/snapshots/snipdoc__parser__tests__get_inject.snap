---
source: snipdoc/src/parser/mod.rs
expression: parser.inject(&snippet_refs)
---
Ok(
    InjectSummary {
        content: "# Snipdoc\n\n<!-- <snip id=\"installation\" inject_from=\"code\"> -->\n$ cargo install snipdoc\n$ snipdoc --version\n<!-- </snip> -->\n\n<!-- <snip id=\"inject_from_yaml\" inject_from=\"yaml\"> -->\ninject_from_yaml\n<!-- </snip> -->\n\n<!-- <snip id=\"inject_from_yaml\" inject_from=\"code\"> -->\n# Skip injection, `inject_from_yaml` snippet id not exists in code\n<!-- </snip> -->\n\n<!-- <snip id=\"inject_from_yaml\" inject_from=\"any\"> -->\ninject_from_yaml\n<!-- </snip> -->\n\n<!-- <snip id=\"description\" inject_from=\"code\" add_prefix=\"//! \"> -->\n//! snipdoc\n<!-- </snip> -->\n\n<!-- <snip id=\"description\" inject_from=\"code\" strip_prefix=\"snip\"> -->\ndoc\n<!-- </snip> -->\n\n<!-- <snip id=\"description\" inject_from=\"code\" template=\"```sh\\n{snippet}\\n```\"> -->\n```sh\nsnipdoc\n```\n<!-- </snip> -->\n\n<!-- <snip id=\"description\" inject_from=\"code\"> -->\nsnipdoc\n<!-- </snip> -->\n\n<!-- <snip id=\"not-found\" inject_from=\"code\"> -->\nnot-found\n<!-- </snip> -->\n\n",
        actions: [
            Injected {
                snippet_id: "installation",
                content: "$ cargo install snipdoc\n$ snipdoc --version",
            },
            Injected {
                snippet_id: "inject_from_yaml",
                content: "inject_from_yaml",
            },
            Injected {
                snippet_id: "inject_from_yaml",
                content: "inject_from_yaml",
            },
            Injected {
                snippet_id: "description",
                content: "//! snipdoc",
            },
            Injected {
                snippet_id: "description",
                content: "doc",
            },
            Injected {
                snippet_id: "description",
                content: "```sh\nsnipdoc\n```",
            },
            Equal {
                snippet_id: "description",
            },
            NotFound {
                snippet_id: "not-found",
                snippet_kind: Code,
            },
        ],
    },
)
